src/recipe_parser.zig

```diff
@@ -0,0 +1,264 @@
+		for(tags.items) |tag| {
+			if(!item.hasTag(tag) and !(item.block() != null and (Block{.typ = item.block().?, .data = 0}).hasTag(tag))) return itemPairs;
+		}
+		itemPairs.append(.{
+			.item = .{
+				.item = .{.baseItem = item},
+				.amount = amount,
+			},
+			.keys = keys.clone() catch unreachable,
+		});
+	} else {
+		if(std.mem.eql(u8, "{mod}:branch/leafy/{type}", id)) {
+			var iter = keys.iterator();
+			while(iter.next()) |item| {
+				std.log.info("{s}: {s}", .{item.key_ptr.*, item.value_ptr.*});
+			}
+		}
+
+		var iter = items.iterator();
+		loop: while(iter.next()) |item| {
+			for(tags.items) |tag| {
+				if(!item.hasTag(tag) and !(item.block() != null and (Block{.typ = item.block().?, .data = 0}).hasTag(tag))) {
+					continue :loop;
+				}
+			}
+			if(matchWithKeys(allocator, item.id(), pattern.items, keys)) |newKeys| {
+				itemPairs.append(.{
+					.item = .{
+						.item = .{.baseItem = item.*},
+						.amount = amount,
+					},
+					.keys = newKeys,
+				});
+			} else if(id.len == 0) {
+				itemPairs.append(.{
+					.item = .{
+						.item = .{.baseItem = item.*},
+						.amount = amount,
+					},
+					.keys = keys.clone() catch unreachable,
+				});
+			}
+		}
+	}
+	return itemPairs;
+}
+
+fn generateItemCombos(allocator: NeverFailingAllocator, recipe: []ZonElement) !main.List([]ItemStack) {
+	var remainingItems = recipe;
+	var emptyKeys: std.StringHashMap([]const u8) = .init(allocator.allocator);
+	defer emptyKeys.deinit();
+	const startingParsedItems = try parseRecipeItem(allocator, remainingItems[0], &emptyKeys);
+	defer startingParsedItems.deinit();
+	var inputCombos: main.List([]ItemStack) = .initCapacity(allocator, startingParsedItems.items.len);
+	errdefer {
+		for(inputCombos.items) |combo| {
+			allocator.free(combo);
+		}
+		inputCombos.deinit();
+	}
+	var keyList: main.List(std.StringHashMap([]const u8)) = .initCapacity(allocator, startingParsedItems.items.len);
+	defer {
+		for(keyList.items) |*keys| {
+			keys.deinit();
+		}
+		keyList.deinit();
+	}
+	for(startingParsedItems.items) |item| {
+		const inputs = allocator.alloc(ItemStack, recipe.len);
+		inputs[0] = item.item;
+		inputCombos.append(inputs);
+		keyList.append(item.keys);
+	}
+	while(remainingItems.len > 1) {
+		remainingItems = remainingItems[1..];
+		const startIndex = inputCombos.items[0].len - remainingItems.len;
+		var newKeyList: main.List(std.StringHashMap([]const u8)) = .init(allocator);
+		var newInputCombos: main.List([]ItemStack) = .init(allocator);
+
+		for(keyList.items, inputCombos.items) |*keys, inputs| {
+			const parsedItems = try parseRecipeItem(allocator, remainingItems[0], keys);
+			defer parsedItems.deinit();
+			for(parsedItems.items) |item| {
+				const newInputs = allocator.dupe(ItemStack, inputs);
+				newInputs[startIndex] = item.item;
+				newInputCombos.append(newInputs);
+				newKeyList.append(item.keys);
+			}
+		}
+
+		for(keyList.items) |*keys| {
+			keys.deinit();
```

Comment:
deinit should always be done with a defer, before the loop (right after the list creation). This is currently leaking memory on error.

Reply:
same below